app/services/katello/pulp3/repository.rb

```diff
@@ -0,0 +1,152 @@
+        pulp3_api.repositories_list(args).results
+      end
+
+      def delete(href = repository_reference.repository_href)
+        pulp3_api.repositories_delete(href)
+      end
+
+      def update_distribution(path)
+        distribution_reference = distribution_reference(path)
+        pulp3_api.distributions_partial_update(distribution_reference.href, publication: repo.publication_href)
+      end
+
+      def refresh_distributions
+        paths.each do |prefix, path|
+          if distribution_reference(path)
+            update_distribution(path)
+          else
+            create_distribution(prefix, path)
+          end
+        end
+      end
+
+      def create_version
+        #TODO: https://pulp.plan.io/issues/4447
+        pulp3_api.repositories_versions_create(repository_reference.repository_href.split('/').last, {})
+      end
+
+      def paths
+        list = {
+          https: "https/#{repo.relative_path}"
+        }
+        list['http'] = "http/#{repo.relative_path}" if repo.root.unprotected
+        list
+      end
+
+      private def delete_distribution(href)
+        pulp3_api.distributions_delete(href)
+      end
+
+      def lookup_distributions(args)
+        pulp3_api.distributions_list(args).results
+      end
+
+      def delete_distributions_by_paths
+        paths.values.each do |path|
+          dists = lookup_distributions(base_path: path.sub(/^\//, ''))
+          delete_distribution(dists.first._href) if dists.first
+        end
+      end
+
+      def create_distribution(prefix, path)
+        path = path.sub(/^\//, '') #remove leading / if present
+        distribution = pulp3_api.distributions_create(base_path: path, publication: repo.publication_href, name: "#{prefix}_#{backend_object_name}")
+        DistributionReference.create!(path: path, href: distribution._href, root_repository_id: repo.root.id)
+      end
+
+      def common_remote_options
+        remote_options = {
+          validate: true,
+          ssl_validation: root.verify_ssl_on_sync,
+          name: backend_object_name,
+          url: root.url
+        }
+        if root.upstream_username && root.upstream_password
+          remote_options.merge(username: root.upstream_username,
+                               password: root.upstream_password)
+        end
+        remote_options.merge(ssl_remote_options)
+      end
+
+      def ssl_remote_options
+        if root.redhat? && Katello::Resources::CDN::CdnResource.redhat_cdn?(root.url)
+          {
+            ssl_client_certificate: root.product.certificate,
+            ssl_client_key: root.product.key,
+            ssl_ca_certificate: Katello::Repository.feed_ca_cert(root.url)
+          }
+        elsif root.custom?
+          {
+            ssl_client_certificate: root.ssl_client_cert&.content,
+            ssl_client_key: root.ssl_client_key&.content,
+            ssl_ca_certificate: root.ssl_ca_cert&.content
+          }
+        else
+          {}
+        end
```

Comment:
One thing to ponder, should we pull distribution/remote/publisher out into their own classes?   I'm leaning that way, but might make sense in a future pr rather than this one